src/Microsoft.AspNet.Server.Kestrel/Http/SocketOutput.cs

```diff
@@ -388,7 +438,23 @@ void ISocketOutput.Write(ArraySegment<byte> buffer, bool immediate, bool chunk)
 
         Task ISocketOutput.WriteAsync(ArraySegment<byte> buffer, bool immediate, bool chunk, CancellationToken cancellationToken)
         {
-            return WriteAsync(buffer, immediate, chunk);
+            if (cancellationToken.IsCancellationRequested)
+            {
+#if DOTNET5_4
+                return Task.FromCanceled(cancellationToken);
+#else
+                throw new TaskCanceledException();
```

Comment:
I'll have a dig; might be able to do something odd for 451 - like return a faulting lambda task